File: backend/src/main/java/com/languagelearning/service/admin/BadgeManagementService.java
package com.languagelearning.service.admin;

import com.languagelearning.dto.admin.badge_management.BadgeDto;
import com.languagelearning.dto.admin.badge_management.BadgeStatItemDto;
import com.languagelearning.dto.admin.badge_management.BadgeStatsDto;
import com.languagelearning.entity.Badge;
import com.languagelearning.entity.UserBadge;
import com.languagelearning.repository.mysql.BadgeRepository;
import com.languagelearning.repository.mysql.UserBadgeRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class BadgeManagementService {

    private final BadgeRepository badgeRepository;
    private final UserBadgeRepository userBadgeRepository;

    @Transactional(readOnly = true)
    public Page<BadgeDto> getBadges(int page, int size, String keyword) {
        int safePage = Math.max(page, 0);
        int safeSize = Math.max(size, 1);
        String q = keyword == null ? "" : keyword.trim().toLowerCase(Locale.ROOT);

        Map<Integer, Long> recipientCounts = getRecipientCounts();

        List<Badge> filtered = badgeRepository.findAll().stream()
                .filter(badge -> q.isBlank()
                        || containsIgnoreCase(badge.getBadgeName(), q)
                        || containsIgnoreCase(badge.getDescription(), q))
                .sorted(Comparator
                        .comparing((Badge badge) -> badge.getRequiredKn() == null ? 0 : badge.getRequiredKn())
                        .thenComparing(Badge::getId, Comparator.nullsLast(Integer::compareTo)))
                .toList();

        int start = Math.min(safePage * safeSize, filtered.size());
        int end = Math.min(start + safeSize, filtered.size());
        List<BadgeDto> content = filtered.subList(start, end).stream()
                .map(badge -> toDto(badge, recipientCounts.getOrDefault(badge.getId(), 0L)))
                .toList();

        return new PageImpl<>(content, PageRequest.of(safePage, safeSize), filtered.size());
    }

    @Transactional(readOnly = true)
    public BadgeStatsDto getStats() {
        List<Badge> badges = badgeRepository.findAll().stream()
                .sorted(Comparator
                        .comparing((Badge badge) -> badge.getRequiredKn() == null ? 0 : badge.getRequiredKn())
                        .thenComparing(Badge::getId, Comparator.nullsLast(Integer::compareTo)))
                .toList();

        Map<Integer, Long> recipientCounts = getRecipientCounts();
        List<UserBadge> userBadges = userBadgeRepository.findAll();

        long totalAwards = userBadges.size();
        long uniqueEarners = userBadges.stream()
                .map(UserBadge::getUser)
                .filter(Objects::nonNull)
                .map(user -> user.getId())
                .distinct()
                .count();

        double averageRequiredKn = badges.isEmpty()
                ? 0
                : badges.stream().mapToInt(badge -> badge.getRequiredKn() == null ? 0 : badge.getRequiredKn()).average().orElse(0);

        Integer minRequiredKn = badges.stream()
                .map(Badge::getRequiredKn)
                .filter(Objects::nonNull)
                .min(Integer::compareTo)
                .orElse(null);

        Integer maxRequiredKn = badges.stream()
                .map(Badge::getRequiredKn)
                .filter(Objects::nonNull)
                .max(Integer::compareTo)
                .orElse(null);

        Badge topBadge = badges.stream()
                .max(Comparator.comparingLong(badge -> recipientCounts.getOrDefault(badge.getId(), 0L)))
                .orElse(null);

        List<BadgeStatItemDto> usage = badges.stream()
                .map(badge -> {
                    long recipients = recipientCounts.getOrDefault(badge.getId(), 0L);
                    double share = totalAwards == 0 ? 0 : (recipients * 100.0) / totalAwards;
                    return BadgeStatItemDto.builder()
                            .id(badge.getId())
                            .badgeName(badge.getBadgeName())
                            .requiredKn(badge.getRequiredKn())
                            .recipientCount(recipients)
                            .recipientShare(share)
                            .build();
                })
                .toList();

        return BadgeStatsDto.builder()
                .totalBadges(badges.size())
                .totalAwards(totalAwards)
                .uniqueEarners(uniqueEarners)
                .averageRequiredKn(averageRequiredKn)
                .minRequiredKn(minRequiredKn)
                .maxRequiredKn(maxRequiredKn)
                .topBadgeName(topBadge != null ? topBadge.getBadgeName() : null)
                .topBadgeRecipients(topBadge == null ? 0 : recipientCounts.getOrDefault(topBadge.getId(), 0L))
                .badgeUsage(usage)
                .build();
    }

    private BadgeDto toDto(Badge badge, long recipientCount) {
        return BadgeDto.builder()
                .id(badge.getId())
                .badgeName(badge.getBadgeName())
                .description(badge.getDescription())
                .requiredKn(badge.getRequiredKn())
                .iconUrl(badge.getIconUrl())
                .recipientCount(recipientCount)
                .build();
    }

    private Map<Integer, Long> getRecipientCounts() {
        return userBadgeRepository.findAll().stream()
                .filter(userBadge -> userBadge.getBadge() != null && userBadge.getBadge().getId() != null)
                .collect(Collectors.groupingBy(userBadge -> userBadge.getBadge().getId(), Collectors.counting()));
    }

    private boolean containsIgnoreCase(String value, String keyword) {
        return value != null && value.toLowerCase(Locale.ROOT).contains(keyword);
    }
}

File: frontend/src/services/admin/badgeManagementService.ts
import apiClient from "@/config/api.ts";

interface ApiResponse<T> {
    success: boolean;
    message: string;
    data: T;
}

interface PageResponse<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    number: number;
    size: number;
}

export interface AdminBadge {
    id: number;
    badgeName: string;
    description: string | null;
    requiredKn: number;
    iconUrl: string | null;
    status: "active" | "inactive";
    recipientCount: number;
}

export interface BadgeUsageStat {
    id: number;
    badgeName: string;
    requiredKn: number | null;
    recipientCount: number;
    recipientShare: number;
}

export interface BadgeStats {
    totalBadges: number;
    totalAwards: number;
    uniqueEarners: number;
    averageRequiredKn: number;
    minRequiredKn: number | null;
    maxRequiredKn: number | null;
    topBadgeName: string | null;
    topBadgeRecipients: number;
    badgeUsage: BadgeUsageStat[];
}

export interface BadgeUpsertPayload {
    badgeName: string;
    description: string;
    requiredKn: number;
    status: "active" | "inactive";
    file?: File | null;
    iconUrl?: string | null;
}

function mapBadge(badge: AdminBadge): AdminBadge {
    return {
        ...badge,
        description: badge.description?.trim() ? badge.description : null,
        iconUrl: badge.iconUrl?.trim() ? badge.iconUrl : null,
        status: badge.status === "inactive" ? "inactive" : "active",
    };
}

function buildBadgeFormData(payload: BadgeUpsertPayload): FormData {
    const formData = new FormData();
    formData.append("badgeName", payload.badgeName);
    formData.append("description", payload.description);
    formData.append("requiredKn", String(payload.requiredKn));
    formData.append("status", payload.status);

    if (payload.iconUrl) {
        formData.append("iconUrl", payload.iconUrl);
    }

    if (payload.file) {
        formData.append("file", payload.file);
    }

    return formData;
}

export const badgeManagementService = {
    async getBadges(page = 0, size = 8, keyword = ""): Promise<{ badges: AdminBadge[]; total: number; totalPages: number; page: number; size: number }> {
        const res = await apiClient.get<ApiResponse<PageResponse<AdminBadge>>>("/admin/badges", {
            params: { page, size, keyword: keyword.trim() || undefined },
        });
        const data = res.data.data;
        return {
            badges: data.content.map(mapBadge),
            total: data.totalElements,
            totalPages: data.totalPages,
            page: data.number,
            size: data.size,
        };
    },

    async getStats(): Promise<BadgeStats> {
        const res = await apiClient.get<ApiResponse<BadgeStats>>("/admin/badges/stats");
        return res.data.data;
    },

    async uploadBadgeImage(file: File): Promise<string> {
        const formData = new FormData();
        formData.append("file", file);

        const res = await apiClient.post<ApiResponse<string>>("/admin/badges/image", formData, {
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });

        return res.data.data;
    },

    async createBadge(payload: BadgeUpsertPayload): Promise<AdminBadge> {
        const res = await apiClient.post<ApiResponse<AdminBadge>>("/admin/badges", buildBadgeFormData(payload), {
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });

        return mapBadge(res.data.data);
    },

    async updateBadge(id: number, payload: BadgeUpsertPayload): Promise<AdminBadge> {
        const res = await apiClient.put<ApiResponse<AdminBadge>>(`/admin/badges/${id}`, buildBadgeFormData(payload), {
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });

        return mapBadge(res.data.data);
    },
};